File: lib/objc/DNObjectiveCConverter.js
import { InputStream, CommonTokenStream, tree as _tree } from 'antlr4'
import ObjectiveCLexer from '../../parser/objc/ObjectiveCLexer.js'
import ObjectiveCParser from '../../parser/objc/ObjectiveCParser.js'
import DNObjectiveCParserListener from './DNObjectiveCParserListener.js'
import { parentPort, workerData } from 'worker_threads'

export default class DNObjectiveCConverter {
    convert(content, cb) {
        if (!content) {
            if (!workerData) {
                if (cb) {
                    cb(null, null)
                }
                return
            }
            content = workerData.content
        }
        if (!cb) {
            cb = this.callback
        }

        try {
            const chars = new InputStream(content)
            const lexer = new ObjectiveCLexer(chars)

            const tokens = new CommonTokenStream(lexer)
            const parser = new ObjectiveCParser(tokens)
            const tree = parser.translationUnit()
            let isPartOfFile = workerData ? workerData.isPartOfFile : false
            const listener = new DNObjectiveCParserListener(cb, isPartOfFile)
            _tree.ParseTreeWalker.DEFAULT.walk(listener, tree)
        } catch (e) {
            cb(null, e)
        }
    }

    callback(result, error) {
        // Send a message to the main thread.
        parentPort.postMessage({ result: result, error: error });
    }
}

new DNObjectiveCConverter().convert()

File: lib/src/objc/dn_objectivec_converter.dart
import 'dart:async';

import 'package:antlr4/antlr4.dart';
import 'package:dart_native_codegen/src/common.dart';
import 'package:dart_native_codegen/src/objc/dn_objectivec_parser_listener.dart';
import 'package:dart_native_codegen/parser/objc/ObjectiveCLexer.dart';
import 'package:dart_native_codegen/parser/objc/ObjectiveCParser.dart';

class DNObjectiveCConverter {
  //convert
  static Future<String> convert(String content) async {
    Completer<String> completer = Completer();
    try {
      final chars = InputStream.fromString(content);
      final lexer = ObjectiveCLexer(chars);
      final tokens = CommonTokenStream(lexer);
      final parser = ObjectiveCParser(tokens);
      parser.addErrorListener(DiagnosticErrorListener());
      parser.buildParseTree = true;
      final tree = parser.translationUnit();
      Callback cb = (String content, {String error}) {
        if (content != null) {
          completer.complete(content);
        } else {
          completer.completeError(error, StackTrace.current);
        }
      };
      ParseTreeWalker.DEFAULT.walk(DNObjectiveCParserListener(cb), tree);
    } catch (e) {
      completer.completeError(e, StackTrace.current);
    }
    return completer.future;
  }
}
